### PR TITLE
Documentation site build rework

### LINKED ISSUE #193: BUG: CI does not capture certain failures
On some occasions (such as [here](https://github.com/SNL-WaterPower/WecOptTool/actions/runs/3840956795/jobs/6540619364), which caused #182), the GitHub release workflow encounters a fatal error, but does not cause a CI failure.

### LINKED ISSUE #203: Feature request: Windows support for Sphinx documentation build
**Feature description.**
Either (a) modify the "Editing the documentation" section in the [Contributing guide](https://github.com/SNL-WaterPower/WecOptTool/blob/main/.github/CONTRIBUTING.md#editing-the-documentation) to provide guidance for Windows users, and/or (b) modify the [makefile](https://github.com/SNL-WaterPower/WecOptTool/blob/main/docs/Makefile) to be compatible with [nmake](https://learn.microsoft.com/en-us/cpp/build/reference/nmake-reference?view=msvc-170).

**Issue addressed.**
The current guidance for building the documentation locally does not work with Windows, as `make` is a Unix-specific function and the makefile includes Unix-specific commands. While this is not a major concern at the moment since the local documentation build is a more esoteric development tool (and the CI builds the documentation on Linux anyway), future contributors developing on Windows may be misguided by the instructions.

**Describe the solution you'd like**
The easiest short-term workaround for this is to modify CONTRIBUTING.md with guidance that Windows users will need to use some emulated Linux environment in order to build the documentation locally; I personally have [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/install) set up, but [cygwin](https://www.cygwin.com/index.html) would also work.

Windows is able to natively build makefiles using the [nmake](https://learn.microsoft.com/en-us/cpp/build/reference/nmake-reference?view=msvc-17) command in Visual Studio, but this requires somewhat different syntax than GNUMake. The longer term solution would be to modify the makefile to [identify if it is operating in a Windows environment](https://stackoverflow.com/questions/4058840/makefile-that-distinguishes-between-windows-and-unix-like-systems) and provide the nmake build commands if it is. This is probably a low priority though. 

**Describe alternatives you've considered**
[chocolatey](https://chocolatey.org/) also ships the make function, but I have encountered errors trying to install Miniconda3 in it with no clear reason, and many operations also require an elevated PowerShell command prompt, which is a bit of a pain. At the very least, my personal experience with Windows Subsystem for Linux is much more straightforward (and that includes dealing with SNL proxies ;) ) and offers more robust functionality.

**Interest in leading this feature development?**
I can lead this.

### PR DESCRIPTION
**Pull request by michaelcdevin**
## Description
WecOptTool uses Sphinx to build documentation via a [Makefile](https://github.com/SNL-WaterPower/WecOptTool/blob/42a27d882a48ff6870e5600dca1263d577641a36/docs/Makefile), but Sphinx handles exceptions with internal try/except clauses that do not return error codes to higher level programs, see https://stackoverflow.com/a/58398714. This prevents Sphinx build errors from being returned to the GitHub Actions CI in the WecOptTool [release workflow](https://github.com/SNL-WaterPower/WecOptTool/blob/main/.github/workflows/release.yml), which can cause website build failures without any notification. This PR replaces the Sphinx build Makefile with `build_docs.py`, which interfaces with Sphinx objects directly, allowing Sphinx build exceptions to propagate to the CI correctly. The Contributing documentation is also updated accordingly.

Fixes #193. Since everything regarding Sphinx is now in pure Python and not dependent on the `make` command, this also conveniently fixes #203.

## Type of PR
- [x] Bug fix
- [x] New feature
- [x] Documentation

## Checklist for PR
- [x] Authors read the [contribution guidelines](https://github.com/SNL-WaterPower/WecOptTool/blob/main/.github/CONTRIBUTING.md)
- [x] The pull request is from an issue branch (not main) on *your* fork, to the [main branch in WecOptTool](https://github.com/SNL-WaterPower/WecOptTool).
- [x] The authors have given the admins edit access
- [x] All changes adhere to the style guide including PEP8, Docstrings, and Type Hints.
- [x] Modified the documentation if applicable
- [x] All tests pass
- [x] [Reference or close any relevant issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)

## Additional details
I tried to translate everything from the old Makefile into either `build_docs.py` or `clean_docs.py`, but I'm no expert in `make` commands. If there's any functionality being lost, let me know and I can add it to this PR.

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -66,15 +66,15 @@ The homepage source code is in `./docs/source/index.rst`.
...
+python3 docs/build_docs.py
 ```
 
 The built documentation will be in `./docs/_build` and the homepage is `./docs/_build/index.html`.
-To delete do `make clean`.
+To delete, do `python3 docs/clean_docs.py`.
 
 The documentation uses the Jupyter notebook tutorials in the `examples` directory.
 When building the documentation locally you will need to have installed [pandoc](https://pandoc.org/installing.html) and [gifsicle](https://github.com/kohler/gifsicle).
+We recommend installing pandoc using its Anaconda distribution: `conda install -c conda-forge pandoc`.
```
Comment (michaelcdevin) [inline .github/CONTRIBUTING.md:77]: Added this while investigating #203. Pandoc is a pain to get playing nice with Python on Windows using the methods suggested on the website, but there is an undocumented precompiled version supported on conda-forge that works like a charm.

File: docs/build_docs.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    app.build()
+
+
+def html():
+    app = Sphinx(source_dir,
+                 conf_dir,
+                 build_dir,
+                 doctree_dir,
+                 'html',
+                 warningiserror=True)
```
Comment (michaelcdevin) [inline docs/build_docs.py:31]: I think writing it like this is redundant, but Sphinx doesn't have a good way of handling multiple builders outside of the Makefile interface at the moment. However, this should be addressed in https://github.com/sphinx-doc/sphinx/issues/5618 whenever it gets merged.

File: docs/source/conf.py [hunk 3]
```diff
@@ -52,10 +52,14 @@
...
 
-suppress_warnings = ["autosectionlabel.*"] # nbsphinx and austosectionlabel do not play well together
+suppress_warnings = ['autosectionlabel.*', # nbsphinx and austosectionlabel do not play well together
+                     'app.add_node', # using multiple builders in custom Sphinx objects throws a bunch of these
+                     'app.add_directive',
+                     'app.add_role',]
 
 linkcheck_ignore = [
     'https://github.com/HIPS/autograd/blob/master/docs/tutorial.md#supported-and-unsupported-parts-of-numpyscipy',
+    'https://doi.org/10.2172/1330189',
```
Comment (michaelcdevin) [inline docs/source/conf.py:62]: Turns out we were getting a warning for this URL in `linkcheck` that was naturally never caught in the CI. See https://github.com/urllib3/urllib3/issues/2653#issuecomment-1165307051 for why this happens. I think we can safely ignore this warning -- the SSLError at the root of this problem should supposedly be solved in Python 3.12 anyway.

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -66,15 +66,15 @@ The homepage source code is in `./docs/source/index.rst`.
 To build the documentation locally (not required but good check)
 
 ```bash
-cd docs
-make html
+python3 docs/build_docs.py
 ```
```
Comment (cmichelenstrofer) [inline .github/CONTRIBUTING.md:70]: the `make html/pdf/etc` and `make clean` are common interfaces. It is a shame to have to replace them. But that's OK, it is not like we use this often. It is mostly just the CI.

**Review by cmichelenstrofer — APPROVED** [review]
Awesome! 👏  :)

**Pull request merged**